Switch to push

mattjgalloway opened the pull request:
This migrates the whole integration to use the STOMP over Websocket connection to Mixergy to support pushing of data rather than regular polling.

@tomasmcguinness: I wanted to put this up now, but I don't want to merge it just yet (so putting it up as a draft). I'm going to play around with it and keep testing it out. It all seems to work fine, but it's a huge change so I want to make sure I haven't missed anything. Also if you're keen to test it out, that would be amazing too.

(Fixes #48)

mattjgalloway committed:
Switch to push

This migrates the whole integration to use the STOMP over Websocket connection to Mixergy to support pushing of data rather than regular polling.

mattjgalloway commented:
I've given this more testing and all seems to be well. It's pretty cool seeing the updates happen in realtime for the tank settings. And measurement data comes in straight away when the tank updates. So I would say we're going to be causing less load for Mixergy with this because it sends us data every minute rather than us polling every 30 seconds.

By the way - I lodged a request to Mixergy to support MQTT in the tank. It's probably a long shot but it would be amazing to have local MQTT support. I want it mainly for very fast updating of the clamp and PV divert power data. The minute by minute data is kind of sucky for graphs. But nothing we can do about that with the data we currently get from Mixergy.

mattjgalloway committed:
Add "maintain target temperature" switch

This adds a switch for the missing "maintain target temperature" setting. I missed this when adding a bunch of these settings before.

tomasmcguinness commented:
I've been asking them annually for local control via an API or MQTT. They keep saying "it's something we want to do".

Would you mind if I pre-released this?

mattjgalloway commented:
> I've been asking them annually for local control via an API or MQTT. They keep saying "it's something we want to do".

How hopeful are you that they'll do it? I really hope they do!

> Would you mind if I pre-released this?

Sure thing! How would you do that?

mattjgalloway commented:
I’m hoping they allow MQTT as a purely developer thing. Not something that’s required to be used. But it probably is a long shot you’re right.

For the branch - you could clone my branch, or give me access to push a branch?

mattjgalloway commented:
I've had this running for a week now on my testing instance. All seems to be going well. I'd consider this stable at this point!

mattjgalloway committed:
Clear stomp connection on disconnect

mattjgalloway committed:
Only try connecting STOMP for 1 minute

I hit a strange problem once when connecting to the Mixergy STOMP server where it never connected, and we sat waiting forever in the `connect` call. So instead, we can not wait while calling `connect` and instead wait ourselves, with a timeout. If that fails, we'll just retry again.

mattjgalloway commented:
I would hold off on this for now. I've found a slight hiccup where the stomp connection isn't properly reconnecting when the internet goes down. I'm doing some more work on this to see how I can make it resilient to internet outage.

mattjgalloway committed:
Fetch data when starting STOMP connection

mattjgalloway committed:
Catch all exceptions and log

mattjgalloway committed:
Reset the right variable

tomasmcguinness commented:
I need to get back to this!

Do you kno if it's possible to merge this PR into another branch in this repo?

mattjgalloway commented:
Sorry yes, I need to get back to this too! It's close, I just wanted to figure out why it locks up when network disconnects.

I have been quite busy with work recently, but I have a whole load of time off coming up, and am planning to dedicate some to many home automation projects! I'll definitely be trying to get this one over the line!

creednmd commented:
Has there been any update on this push? Would like to get this push version installed, but doesn't appear to have all the current entities in that the latest version does? (HA beginner, so unsure how to tweak current code to get it working myself.)

creednmd commented:
Hi @mattjgalloway Cool. That would be great if you can get it both pushing and with all the latest entities as well. Anything I can get working to justify the screen in the hallway and HA system to the significant-other helps. :)

mattjgalloway committed:
Keep a single connection

mattjgalloway committed:
Fix hang when internet drops during push reconnect

mattjgalloway committed:
Fix energy sensor constructors after rebase
